docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+   oe_get_evidence(&format_local_only, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+   oe_get_evidence(&format_remote_only, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, NULL, 0, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+
+   // New
+   oe_get_evidence(&format_sgx_local, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx_remote, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_remote_only, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+User Experience
+---------------
+The user experience is improved by the changes. These are a couple of possible attestation scenarios:
+
+Scenario 1 - Attester selects from a list of evidence formats that can be verified:
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
+    2. Builds the format IDs of the registered verifiers sorted by preference.
+    3. Sends the format IDs to the attester application. 
+- Attester application
+    1. Upon initialization, queries available attesters to the application (oe_sgx_get_attesters) and registers them (oe_register_attester).
+    2. Receives format IDs from the verifier application.
+    3. Selects a format ID from the received format IDs (oe_select_attester_format).
+    4. Sends the format ID to the verifier application.
+- Verifier application
+    1. Gets evidence parameters for the received format ID (oe_get_evidence_parameters).
+    2. Sends the evidence parameters to the attester application.
+- Attester application
+    1. Generates an evidence (oe_get_evidence) of the selected format with the received evidence parameters.
+    2. Sends the evidence to the verifier application.
+- Verifier application
+    1. Verifies the received evidence (oe_verify_evidence).
+
+Scenario 2 - Attester sends evidences of different formats until the verifier can verify:
+- Attester application
+    1. Upon initialization, queries available attesters to the application (oe_sgx_get_attesters) and registers them (oe_register_attester).
+    2. Builds the format IDs of registered attester sorted by perference.
+    3. Generates an evidence (oe_get_evidence) of the first format ID from the list built in step 2.
+    4. Sends the evidence to the verifier application.
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
+    2. Verifies the received evidence (oe_verify_evidence). If the verification fails, ask the attester applicaiton to generate an evidence of different format.
```

Comment:
typo "applicaiton".
However, what is the real world scenario where an attester would ever send an unsupported format to a verifier?

Reply:
Thanks.

---
In reply to: [393255837](https://github.com/openenclave/openenclave/pull/2621#discussion_r393255837) [](ancestors = 393255837)